source/arm9/libc/fatfs/cache.c

```diff
@@ -22,33 +23,48 @@ static cache_entry_t *cache_entries;
 static uint8_t *cache_mem;
 static uint32_t cache_num_sectors;
 
-int cache_init(uint32_t num_sectors)
+extern uint8_t* dldiGetStubDataEnd(void);
+extern uint8_t* dldiGetStubEnd(void);
+
+int cache_init(int32_t num_sectors)
 {
     // If this function is called after the first time, clear the cache and
     // allocate a new one.
 
     if (cache_entries != NULL)
         free(cache_entries);
 
-    if (cache_mem != NULL)
+    if (cache_mem != NULL && (uintptr_t)cache_mem > (uintptr_t)dldiGetStubEnd())
         free(cache_mem);
 
-    cache_entries = calloc(num_sectors, sizeof(cache_entry_t));
-    if (cache_entries == NULL)
-        return -1;
+    int32_t stub_space_sectors = (dldiGetStubEnd() - dldiGetStubDataEnd()) >> 9;
+    if (num_sectors < 0) {
+        num_sectors = stub_space_sectors;
+    }
+
+    if (num_sectors > 0) {
+        cache_entries = calloc(num_sectors, sizeof(cache_entry_t));
+        if (cache_entries == NULL)
+            return -1;
 
 #if FF_MAX_SS != FF_MIN_SS
 #error "Set the block size to the right value"
 #endif
 
-    cache_mem = malloc(num_sectors * FF_MAX_SS);
-    if (cache_mem == NULL)
-    {
-        free(cache_entries);
-        return -1;
-    }
+        // If sufficient, use the space at the end of the DLDI stub
+        cache_mem = num_sectors <= stub_space_sectors
+            ? dldiGetStubEnd() - (num_sectors << 9)
+            : malloc(num_sectors * FF_MAX_SS);
```

Comment:
Maybe add a comment here saying that the space allocated by malloc() will never be freed, so this can't cause memory leaks.